[datadog_reference_table] Add terraform support for reference tables

g-brizolier opened the pull request:
## What & Why

Adds Terraform support for Datadog Reference Tables, enabling users to manage Reference Tables as Infrastructure-as-Code.

**Includes:**

- `datadog_reference_table` resource - CRUD operations for Reference Tables
- `datadog_reference_table` data source - Query tables by ID or name
- `datadog_reference_table_rows` data source - Fetch specific rows by ID
- Schema validation via `ModifyPlan` to prevent destructive changes (only additive schema changes allowed)
- Support for cloud storage sources (AWS S3, GCP GCS, Azure Blob)

**Note:** Depends on OpenAPI spec fix ([DataDog/datadog-api-spec#4532](https://github.com/DataDog/datadog-api-spec/pull/4532)) to resolve OneOf unmarshaling for `file_metadata`.

## QA Instructions

### Run Acceptance Tests locally (they also run in CI)

```bash
cd ~/dd/terraform-provider-datadog && go build ./... && eval $(dd-auth --domain dd.datad0g.com --output) \
&& export DD_TEST_CLIENT_API_KEY="$DD_API_KEY" \
&& export DD_TEST_CLIENT_APP_KEY="$DD_APP_KEY" \
&& export DD_TEST_SITE_URL="https://api.datad0g.com" \
&& export DD_HOST="https://api.datad0g.com" \
&& export DD_TEST_ORG="yB5yjZ" \
&& TF_ACC=1 go test -v ./datadog/tests/... -run "TestAccReferenceTable"
```

Tests cover:

- Resource CRUD (create, update, delete)
- Schema validation (additive vs destructive changes)
- Cloud storage sources (S3, GCS, Azure)
- Data source queries (by ID, by name, rows)
- Import functionality

### Manual Testing (Optional)

1. Build provider: `make`
2. Setup `~/.terraformrc` with dev_overrides pointing to `$GOPATH/bin`
3. Create test HCL in a separate directory
4. Use `dd-auth --domain dd.datad0g.com --api-key-env=TF_VAR_datadog_api_key --app-key-env=TF_VAR_datadog_app_key terraform plan`

## References

- [Terraform Provider Development Docs](https://datadoghq.atlassian.net/wiki/spaces/API/pages/3881239106/Getting+Started+-+Terraform+Provider)
- [dd-auth CLI Documentation](https://datadoghq.atlassian.net/wiki/spaces/COBS/pages/4094591351/Datadog+Authentication+CLI+dd-auth)

linus-sun-dd commented:
<img width="565" height="114" alt="image" src="https://github.com/user-attachments/assets/79e76be6-31d9-4b79-b090-d69af0db439b" />

i'm seeing `no tests to run` when executing the acceptance tests, is this expected?

p-garratt commented on docs/data-sources/reference_table.md:
```diff
@@ -0,0 +1,105 @@
+Read-Only:
+
+- `fields` (Block List) List of fields in the table schema. (see [below for nested schema](#nestedblock--schema--fields))
+- `primary_keys` (List of String) List of field names that serve as primary keys for the table.
```
Should probably have some comment about how the primary_key must be in fields. It's pretty obvious, but good to be explicit about how these fields relate.

g-brizolier commented:
Addressed review comments:
- Unified error messaging for reference table lookups
- Added 250 row limit validation with clear error message
- Added ValidateConfig to enforce: exactly one access_details type, matching source type, non-empty schema fields and primary_keys
- Improved readability with local variables for AWS/GCP/Azure details
- Fixed trailing newlines in examples

g-brizolier commented on datadog/fwprovider/data_source_datadog_reference_table.go:
```diff
@@ -0,0 +1,458 @@
+		tableId := ddResp.Data[0].GetId()
+		fullResp, _, err := d.Api.GetTable(d.Auth, tableId)
+		if err != nil {
+			response.Diagnostics.Append(utils.FrameworkErrorDiag(err, "error getting datadog reference table details"))
```
Fixed - unified to 'error getting reference table'

g-brizolier commented on docs/data-sources/reference_table.md:
```diff
@@ -0,0 +1,105 @@
+
+Read-Only:
+
+- `aws_detail` (Block, Read-only) AWS S3 access configuration. (see [below for nested schema](#nestedblock--file_metadata--access_details--aws_detail))
```
Added! The resource description now links to the official docs: https://docs.datadoghq.com/reference_tables/?tab=cloudstorage#create-a-reference-table which includes setup instructions for granting Datadog read access to cloud storage buckets.

g-brizolier commented on examples/resources/datadog_reference_table/azure.tf:
```diff
@@ -0,0 +1,47 @@
+    }
+  }
+
+  tags = ["source:azure", "team:warehouse"]
```
Done - removed tags from azure.tf to show an example without tags.

p-garratt commented on datadog/fwprovider/data_source_datadog_reference_table_rows.go:
```diff
@@ -0,0 +1,176 @@
+	retryErr := retry.RetryContext(ctx, 1*time.Minute, func() *retry.RetryError {
+		ddResp, httpResp, err = d.Api.GetRowsByID(d.Auth, tableId, rowIds)
+		if err != nil {
+			// If we get a 404, the table might not have synced yet - retry
```
No changes needed, but man will it be nice when we have read after write consistency and this is no longer needed.

None of this retry will be needed because we'll get the rows written synchronously too

p-garratt commented on datadog/fwprovider/data_source_datadog_reference_table_rows.go:
```diff
@@ -0,0 +1,165 @@
+			return nil
+		}
+		// If we got some rows but not all, the table might still be syncing - retry
+		if len(ddResp.Data) > 0 && len(ddResp.Data) < len(rowIds) {
```
The customer behavior on partial row matches will be somewhat frustrating, because they will get an error rather than data and they won't know which row ids are wrong.
Is there a way to return the partial rows if we've exhausted the retries?

p-garratt commented on datadog/fwprovider/resource_datadog_reference_table.go:
```diff
@@ -0,0 +1,890 @@
+		schemaChanged := false
+
+		// Compare primary keys
+		var statePKs, planPKs []string
```
Can you just use slices.Equal()?

p-garratt commented on datadog/tests/resource_datadog_reference_table_test.go:
I might of just missed it, but is there a set of tests for the validation logic? Would probably be good to document that.

p-garratt commented on docs/resources/reference_table.md:
```diff
@@ -0,0 +1,214 @@
+
+- `created_by` (String) UUID of the user who created the reference table.
+- `id` (String) The ID of this resource.
+- `last_updated_by` (String) UUID of the user who last updated the reference table.
```
Just out of curiosity, when they use the API key via TF, will this field have the service account or just be empty?
